src/Microsoft.Data.SqlClient/tests/FunctionalTests/SslOverTdsStreamTest.cs

```diff
@@ -0,0 +1,278 @@
+
+        private static void AsyncTest(int encapsulatedPacketCount, int passthroughPacketCount, int maxPacketReadLength)
+        {
+            byte[] input;
+            byte[] output;
+            SetupArrays(encapsulatedPacketCount + passthroughPacketCount, out input, out output);
+            byte[] buffer = WritePackets(encapsulatedPacketCount, passthroughPacketCount,
+                async (Stream stream, int index) =>
+                {
+                    await stream.WriteAsync(input, TdsEnums.DEFAULT_LOGIN_PACKET_SIZE * index, TdsEnums.DEFAULT_LOGIN_PACKET_SIZE);
+                }
+            );
+
+            ReadPackets(buffer, encapsulatedPacketCount, passthroughPacketCount, maxPacketReadLength, output,
+                async (Stream stream, byte[] bytes, int offset, int count) =>
+                {
+                    return await stream.ReadAsync(bytes, offset, count);
+                }
+            );
+
+            Validate(input, output);
+        }
+
+        static partial void AsyncCoreTest(int encapsulatedPacketCount, int passthroughPacketCount, int maxPacketReadLength);
+
+
+
+        private static void ReadPackets(byte[] buffer, int encapsulatedPacketCount, int passthroughPacketCount, int maxPacketReadLength, byte[] output, Func<Stream, byte[], int, int, Task<int>> action)
+        {
+            using (LimitedMemoryStream stream = new LimitedMemoryStream(buffer, maxPacketReadLength))
+            using (Stream tdsStream = CreateSslOverTdsStream(stream))
+            {
+                int offset = 0;
+                byte[] bytes = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+                for (int index = 0; index < encapsulatedPacketCount; index++)
+                {
+                    Array.Clear(bytes, 0, bytes.Length);
+                    int packetBytes = ReadPacket(tdsStream, action, bytes).GetAwaiter().GetResult();
+                    Array.Copy(bytes, 0, output, offset, packetBytes);
+                    offset += packetBytes;
+                }
+                InvokeFinishHandshake(tdsStream);//tdsStream.FinishHandshake();
+                for (int index = 0; index < passthroughPacketCount; index++)
+                {
+                    Array.Clear(bytes, 0, bytes.Length);
+                    int packetBytes = ReadPacket(tdsStream, action, bytes).GetAwaiter().GetResult();
+                    Array.Copy(bytes, 0, output, offset, packetBytes);
+                    offset += packetBytes;
+                }
+            }
+        }
+
+        private static void InvokeFinishHandshake(Stream stream)
+        {
+            MethodInfo method = stream.GetType().GetMethod("FinishHandshake", BindingFlags.Public | BindingFlags.Instance);
+            method.Invoke(stream, null);
+        }
+
+        private static Stream CreateSslOverTdsStream(Stream stream)
+        {
+            Type type = typeof(SqlClientFactory).Assembly.GetType("Microsoft.Data.SqlClient.SNI.SslOverTdsStream");
```

Comment:
Seems like your "Core" reflects Managed SNI, so I guess method naming needs to be done better.

Reply:
SslOverTdsStream currently only exists in core. I built the tests so that if core was built for versions of core without with async stream extensions (which would be 2.1 or lower for core or all netfx) that it could still be used. Say for example you wanted to replace parts of netfx build with parts of the netcore build it would make testing easier for you.

I can rename if you want but I thought common things would be the normal case and that things on netcore only could have the core suffix. Any suggestions on what to change it to?

Reply:
Ok, I see.
I think we should in that case consider adding additional Moniker in Xunit Extensions that can be applied to specific tests. without the need to splitting classes.

@JRahnama can we extend our `TargetFrameworkMonikers` to include "Netcoreapp21" that when set would skip on "NetCoreApp2.1" target framework?

Reply:
It should work on netcore 2.1 according to the api tables so I don't think we need to add more testing capability. I just found it useful to split the functionality into core and netfx files with common parts rather than use #defines so it can be made available on netfx if/when we converge into a single codebase that contains a managed sni implementation.

I'm open to changing the names if you have suggestions.

Reply:
@cheenamalhotra are we still going to proceed?

Reply:
If your tests are compatible with netcoreapp2.1, do we still need to split the test file? Keeping it in 1 file would be better.

Reply:
I explained my thinking above. I think it's ok this way. If you really want I can change it but I don't see it as a problem to have it split apart the way it is, the complexity isn't huge.